Make defunctionalization use type synonyms' return kinds when available

Surprisingly, the following code does not typecheck:

```hs
{-# LANGUAGE DataKinds #-}
{-# LANGUAGE GADTs #-}
{-# LANGUAGE PolyKinds #-}
{-# LANGUAGE TemplateHaskell #-}
{-# LANGUAGE TypeFamilies #-}
{-# OPTIONS_GHC -ddump-splices #-}
module Bug where

import Data.Kind
import Data.Singletons.TH

type family Any :: k
$(singletons [d| type AnyOfKind (k :: Type) = (Any :: k) |])
```
```
$ /opt/ghc/8.6.5/bin/ghc -package-db ~/.cabal/store/ghc-8.6.5/package.db/ Bug.hs
[1 of 1] Compiling Bug              ( Bug.hs, Bug.o )
Bug.hs:13:3-59: Splicing declarations
    singletons
      [d| type AnyOfKind_aaNb (k_abgL :: Type) = (Any :: k_abgL) |]
  ======>
    type AnyOfKind_aej9 (k_aeja :: Type) = (Any :: k_aeja)
    type AnyOfKindSym1 (k6989586621679064796 :: Type) =
        AnyOfKind_aej9 k6989586621679064796
    instance SuppressUnusedWarnings AnyOfKindSym0 where
      suppressUnusedWarnings = snd (((,) AnyOfKindSym0KindInference) ())
    data AnyOfKindSym0 k6989586621679064796
      where
        AnyOfKindSym0KindInference :: forall k6989586621679064796
                                             arg_aejj. SameKind (Apply AnyOfKindSym0 arg_aejj) (AnyOfKindSym1 arg_aejj) =>
                                      AnyOfKindSym0 k6989586621679064796
    type instance Apply AnyOfKindSym0 k6989586621679064796 = AnyOfKind_aej9 k6989586621679064796

Bug.hs:13:3: error:
    • Type constructor argument ‘k6989586621679064796’ is used dependently.
      Any dependent arguments must be obviously so, not inferred
      by the type-checker.
      Inferred argument kinds: k6989586621679064796 :: *
      Suggestion: use ‘k6989586621679064796’ in a kind to make the dependency clearer.
    • In the type synonym declaration for ‘AnyOfKindSym1’
   |
13 | $(singletons [d| type AnyOfKind (k :: Type) = (Any :: k) |])
   |   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

The culprit lies within the definition of `AnyOfKindSym0`. Curiously, this declaration does not have an explicit kind signature of `:: Type ~> k`, despite the fact that `AnyOfKind`'s definition provides all of that information. In fact, `AnyOfKind` has a CUSK, so it seems especially questionable for `singletons` not to generate defunctionalization symbols with CUSKs as well.

Fixing this would be a simple matter of changing `D.S.Promote.Defun.buildDefunSymsTySynD` to make use of an explicit kind signature on the right-hand side of a type synonym when one is available. Patch incoming.

--- a/src/Data/Singletons/Partition.hs
+++ b/src/Data/Singletons/Partition.hs
@@ -116,9 +116,9 @@ partitionDec (DInstanceD _ _ cxt ty decs) = do
     split_app_tys acc (DSigT t _)   = split_app_tys acc t
     split_app_tys _ _ = fail $ "Illegal instance head: " ++ show ty
 partitionDec (DRoleAnnotD {}) = return mempty  -- ignore these
-partitionDec (DTySynD name tvbs _type) =
+partitionDec (DTySynD name tvbs rhs) =
   -- See Note [Partitioning, type synonyms, and type families]
-  pure $ mempty { pd_ty_syn_decs = [TySynDecl name tvbs] }
+  pure $ mempty { pd_ty_syn_decs = [TySynDecl name tvbs rhs] }
 partitionDec (DClosedTypeFamilyD tf_head _) =
   -- See Note [Partitioning, type synonyms, and type families]
   pure $ mempty { pd_closed_type_family_decs = [TypeFamilyDecl tf_head] }

--- a/src/Data/Singletons/Promote/Defun.hs
+++ b/src/Data/Singletons/Promote/Defun.hs
@@ -42,7 +42,7 @@ defunTypeDecls :: [TySynDecl]
                -> PrM ()
 defunTypeDecls ty_syns c_tyfams o_tyfams = do
   defun_ty_syns <-
-    concatMapM (\(TySynDecl name tvbs) -> buildDefunSymsTySynD name tvbs) ty_syns
+    concatMapM (\(TySynDecl name tvbs rhs) -> buildDefunSymsTySynD name tvbs rhs) ty_syns
   defun_c_tyfams <-
     concatMapM (buildDefunSymsClosedTypeFamilyD . getTypeFamilyDecl) c_tyfams
   defun_o_tyfams <-
@@ -56,8 +56,8 @@ buildDefunSyms (DClosedTypeFamilyD tf_head _) =
   buildDefunSymsClosedTypeFamilyD tf_head
 buildDefunSyms (DOpenTypeFamilyD tf_head) =
   buildDefunSymsOpenTypeFamilyD tf_head
-buildDefunSyms (DTySynD name tvbs _type) =
-  buildDefunSymsTySynD name tvbs
+buildDefunSyms (DTySynD name tvbs rhs) =
+  buildDefunSymsTySynD name tvbs rhs
 buildDefunSyms (DClassD _cxt name tvbs _fundeps _members) = do
   defunReifyFixity name tvbs (Just (DConT constraintName))
 buildDefunSyms _ = fail $ "Defunctionalization symbols can only be built for " ++
@@ -83,9 +83,18 @@ buildDefunSymsTypeFamilyHead default_tvb default_kind
       res_kind = default_kind (resultSigToMaybeKind result_sig)
   defunReifyFixity name arg_tvbs res_kind
 
-buildDefunSymsTySynD :: Name -> [DTyVarBndr] -> PrM [DDec]
-buildDefunSymsTySynD name tvbs =
-  defunReifyFixity name tvbs Nothing
+buildDefunSymsTySynD :: Name -> [DTyVarBndr] -> DType -> PrM [DDec]
+buildDefunSymsTySynD name tvbs rhs =
+  defunReifyFixity name tvbs mb_res_kind
+  where
+    -- In order to have a CUSK, a type synonym must annotate its right-hand
+    -- side with an explicit kind signature, so we can make use of this
+    -- property to determine the result kind when defunctionalizing the
+    -- type synonym.
+    mb_res_kind :: Maybe DKind
+    mb_res_kind = case rhs of
+                    DSigT _ k -> Just k
+                    _         -> Nothing
 
 buildDefunSymsDataD :: [DCon] -> PrM [DDec]
 buildDefunSymsDataD ctors =

--- a/src/Data/Singletons/Syntax.hs
+++ b/src/Data/Singletons/Syntax.hs
@@ -48,7 +48,7 @@ type SingDSigPaInfos = [(DExp, DType)]
 data DataDecl = DataDecl Name [DTyVarBndr] [DCon]
 
 -- The parts of type synonyms that are relevant to singletons.
-data TySynDecl = TySynDecl Name [DTyVarBndr]
+data TySynDecl = TySynDecl Name [DTyVarBndr] DType
 
 -- The parts of open type families that are relevant to singletons.
 type OpenTypeFamilyDecl = TypeFamilyDecl 'Open

--- a/tests/SingletonsTestSuite.hs
+++ b/tests/SingletonsTestSuite.hs
@@ -117,6 +117,7 @@ tests =
     , compileAndDumpStdTest "T367"
     , compileAndDumpStdTest "T371"
     , compileAndDumpStdTest "T376"
+    , compileAndDumpStdTest "T402"
     ],
     testCompileAndDumpGroup "Promote"
     [ compileAndDumpStdTest "Constructors"

--- a/tests/compile-and-dump/Singletons/T402.ghc88.template
+++ b/tests/compile-and-dump/Singletons/T402.ghc88.template
@@ -1,0 +1,15 @@
+Singletons/T402.hs:0:0:: Splicing declarations
+    singletons [d| type AnyOfKind (k :: Type) = Any :: k |]
+  ======>
+    type AnyOfKind (k :: Type) = Any :: k
+    type AnyOfKindSym1 (k0123456789876543210 :: Type) =
+        AnyOfKind k0123456789876543210
+    instance SuppressUnusedWarnings AnyOfKindSym0 where
+      suppressUnusedWarnings = snd (((,) AnyOfKindSym0KindInference) ())
+    data AnyOfKindSym0 :: forall (k0123456789876543210 :: Type).
+                          (~>) Type k0123456789876543210
+      where
+        AnyOfKindSym0KindInference :: forall k0123456789876543210
+                                             arg. SameKind (Apply AnyOfKindSym0 arg) (AnyOfKindSym1 arg) =>
+                                      AnyOfKindSym0 k0123456789876543210
+    type instance Apply AnyOfKindSym0 k0123456789876543210 = AnyOfKind k0123456789876543210

--- a/tests/compile-and-dump/Singletons/T402.hs
+++ b/tests/compile-and-dump/Singletons/T402.hs
@@ -1,0 +1,7 @@
+module T402 where
+
+import Data.Kind
+import Data.Singletons.TH
+
+type family Any :: k
+$(singletons [d| type AnyOfKind (k :: Type) = Any :: k |])

